### PR TITLE
Florian

### PR DESCRIPTION
**Pull request by Florian-collab**
send code changes to the project

### CODE REVIEW

**Comment by Florian-collab** [comment]
I change CGH and CGHerror

File: myCGHError.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+import requests
```
Comment (rtavenar) [inline myCGHError.py:1]: Hum, I guess this file should not be included here. Please remove it.

File: mygraphhv2.py [hunk 2]
```diff
@@ -0,0 +1,531 @@
+import requests
```
Comment (rtavenar) [inline mygraphhv2.py:1]: Here, also, this is a duplicate of `CHG.py`, please include only one of them.

File: graphh/CGH.py [hunk 3]
```diff
@@ -1,45 +1,48 @@
...
-     api_key: str
-        API key to be used for queries
-     premium: bool
-        Whether the account corresponding to this key is a premium account
-        or not
 
+class GraphHopper:
+    """GraphHopper API class.
+
+        Parameters
```
Comment (rtavenar) [inline graphh/CGH.py:14]: Indentation level for the docstring should be that of the beginning of the string (same level as the initial `"""`), cf old version if needed.

File: graphh/CGH.py [hunk 4]
```diff
@@ -1,45 +1,48 @@
...
+        ----------
+         api_key: str
+            API key to be used for queries
+         premium: bool
+            Whether the account corresponding to this key is a premium account
+            or not
+    """
+    """"
```
Comment (rtavenar) [inline graphh/CGH.py:22]: This line is useless (`""""`)

File: graphh/CGH.py [hunk 5]
```diff
@@ -1,45 +1,48 @@
...
+    """GraphHopper API class.
+
+        Parameters
+        ----------
+         api_key: str
+            API key to be used for queries
+         premium: bool
+            Whether the account corresponding to this key is a premium account
+            or not
+    """
```
Comment (rtavenar) [inline graphh/CGH.py:21]: This should not be the end of the docstring (examples are just below), so please do not close it with the `"""`.

File: graphh/CGH.py [hunk 6]
```diff
@@ -48,42 +51,84 @@ class GraphHopper(object):
...
+
         self.api_key = api_key
         self.premium = premium
 
-    def _url_handle(self, api, l_parameters):
-        """This function does an url request with given parameters
+
+
+    def _url_requests(self, api, data, request="get"):
+
```
Comment (rtavenar) [inline graphh/CGH.py:66]: There should be no blank line between the function definition and the docstring (probably necessary for automatic doc generation to work properly).

File: graphh/CGH.py [hunk 7]
```diff
@@ -48,42 +51,84 @@ class GraphHopper(object):
...
 
         Parameters
         ----------
-         api: str
+        api: str
             name of the api used
-         l_parameters: list
-            list of parameters to insert in the url
+        data: dict
+            dict of parameters to insert in the url
```
Comment (rtavenar) [inline graphh/CGH.py:75]: Please add the `request` argument to the docstring and document it properly.

File: graphh/CGH.py [hunk 8]
```diff
@@ -48,42 +51,84 @@ class GraphHopper(object):
...
+            The dictionary return by url request (GET or POST)
+        """
+
         complete_url = GraphHopper.url + api + "?"
-        for p in l_parameters:
-            complete_url += "&{}".format(p)
-        complete_url += "&key=" + self.api_key
+
+        if request.upper() == "POST":
+            complete_url += "key={}".format(self.api_key)
```
Comment (rtavenar) [inline graphh/CGH.py:85]: Is there a reson why the key is not added to the data dict as for the GET query?

File: graphh/CGH.py [hunk 9]
```diff
@@ -48,42 +51,84 @@ class GraphHopper(object):
...
         try:
-            fp = urlopen(complete_url)
-            data = fp.read()
-            encoding = fp.info().get_content_charset('utf-8')
-            result = json.loads(data.decode(encoding))
-        except HTTPError as e:
-            CGHError.CGHError(e)
+            reponse.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+             myCGHError.CGHError(e)
```
Comment (rtavenar) [inline graphh/CGH.py:95]: Please put all your necessary code in `CGHError` and call only functions from this module.

File: graphh/CGH.py [hunk 10]
```diff
@@ -183,94 +235,138 @@ def route(self, l_latlong , vehicle="car", locale="en",
...
+        myCGHError.check_vehicle(vehicle, self.premium)
+        data["vehicle"] = vehicle
+
+        data["out_arrays"] = ["distances", "times", "weights"]
+
+        return self._url_requests("matrix", data, request)
+
+
+
+    def maxtrix(self,l_from_address, l_to_address,out_array, format="default", vehicle="car", request="get"):
```
Comment (rtavenar) [inline graphh/CGH.py:314]: Hum, the name for this method is probably not the one you wanted to use...

File: graphh/CGH.py [hunk 11]
```diff
@@ -183,94 +235,138 @@ def route(self, l_latlong , vehicle="car", locale="en",
...
+        vehicle : str, optional
             The type of vehicle chosen in the list : ["car", "foot", "bike"]
             if the acount is not premium
             And can be chosen in the list : ["small_truck", "truck", "scooter",
             "hike", "mtb", "racingbike"] if it is
-            By default the vehicle will be car
-        unit: str
-            The unit of the distance returned chosen between "ms", "s", "min"
-            and "h"
-            By default the unit will be in milliseconds
```
Comment (rtavenar) [inline graphh/CGH.py:331]: Please document the `format` parameter :)

**Review by rtavenar — COMMENTED** [review]
This is really going in the right direction. My main concern is that you should store all the necessary code in `CGH.py` and `CGHError.py` and remove the other 2 files.

And then I have some formatting concerns that we can deal with step by step through the review process.

Good job, keep going on!

File: graphh/CGH.py (hunk 7 above)
Comment (rtavenar) [reply graphh/CGH.py:71]: There is no `s` at the end of `request`: be careful when you name arguments: what you write here is used to generate the docs.

File: graphh/CGH.py (hunk 9 above)
Comment (rtavenar) [reply graphh/CGH.py:91]: Here you call `myCGHError.CGHError`: check all occurrences of `myCGHError` in your code, they should all be removed.

**Comment by rtavenar** [comment]
Could you please take all my comments from my previous review into account and upload an adapted version? 

Also, as you write it, you code adds a new dependency to the package (`pandas`). It would be better to have something like:

```python
try:
  import pandas
  PANDAS_IMPORTED = True
except ImportError:
  PANDAS_IMPORTED = False

[...]

def function_in_which_I_need_pandas(...):
  if not PANDAS_IMPORTED:
    raise ValueError("pandas was not imported correctly, hence function_in_which_I_need_pandas is not available")
```

**Pull request merged**